### PR TITLE
FIXES #1584

### LINKED ISSUE #1584: Extending NewsController not working
## Bug Report

**Current Behavior**
When installing an extension for example `ext:eventnews`, which extends the `NewsController`, the proxy class is not generated correctly.

Extending the controller:

    $GLOBALS['TYPO3_CONF_VARS']['EXT']['news']['classes']['Controller/NewsController'][] = 'eventnews';

Error:

> ParseError: syntax error, unexpected '$newsRepository' (T_VARIABLE), in file .../public/typo3temp/var/cache/code/news/tx_news_controller_newscontroller.php:824

Looking into the generated class, I see the following (wrong parenthesises):

    public function __construct()
    {
        NewsRepository $newsRepository,
        CategoryRepository $categoryRepository,
        TagRepository $tagRepository,
        ConfigurationManagerInterface $configurationManager
    ) {
        $this->newsRepository = $newsRepository;
        $this->categoryRepository = $categoryRepository;
        $this->tagRepository = $tagRepository;
        $this->configurationManager = $configurationManager;
    }

**Expected behavior/output**
The constructor of the class needs to be generated correctly:

    /**
     * NewsController constructor.
     * @param NewsRepository $newsRepository
     * @param CategoryRepository $categoryRepository
     * @param TagRepository $tagRepository
     * @param ConfigurationManagerInterface $configurationManager
     */
    public function __construct(
        NewsRepository $newsRepository,
        CategoryRepository $categoryRepository,
        TagRepository $tagRepository,
        ConfigurationManagerInterface $configurationManager
    ) {
        $this->newsRepository = $newsRepository;
        $this->categoryRepository = $categoryRepository;
        $this->tagRepository = $tagRepository;
        $this->configurationManager = $configurationManager;
    }

**Environment**
- TYPO3 version(s): 11.4.0
- news version: 9.0.0-dev
- Is your TYPO3 installation set up with Composer (Composer Mode): both

### PR DESCRIPTION
**Pull request by stigfaerch**
[BUGFIX] Generated proxy class is broken

### CODE REVIEW

**Comment by stigfaerch** [comment]
@georgringer 
Had a chance to look into this?

File: Classes/Utility/ClassCacheManager.php [hunk 1]
```diff
@@ -27,11 +27,17 @@ class ClassCacheManager
      */
     protected $classCache;
 
+    /**
+     * @var array
+     */
+    protected $constructorParameters = [];
```
Comment (kitzberger) [inline Classes/Utility/ClassCacheManager.php:33]: @stigfaerch, what's that being used for? Dead code?

File: Classes/Utility/ClassCacheManager.php [hunk 2]
```diff
@@ -162,12 +169,20 @@ protected function parseSingleFile($filePath, $baseClass = false): string
...
-                if (trim($innerPart[$i]) === '{') {
+            $docLineCount = substr_count($constructorInfo['doc'], "\n") + 1;
+            if($baseClass) $this->constructorLines['doc'] = $constructorInfo['doc'];
+            $unsetLine = true;
+            for ($i = ($constructorInfo['start']  - $docLineCount - 1) - ($offsetForInnerPart); $i < $constructorInfo['end'] - $offsetForInnerPart; $i++) {
+                if ($unsetLine) {
+                    if(trim($innerPart[$i]) === ') {' || trim($innerPart[$i]) === '{') {
+                        $unsetLine = false;
+                    } elseif($baseClass && trim($innerPart[$i]) !== ')' && $i > ($constructorInfo['start'] - 1)) {
+                        $this->constructorLines['parameters'][] = LF . $innerPart[$i];
```
Comment (kitzberger) [inline Classes/Utility/ClassCacheManager.php:180]: @stigfaerch, so it's not possible to add other parameters to the constructor? Like EXT:eventnews's `LocationRepository $locationRepository` parameter to the NewsController? https://github.com/georgringer/eventnews/pull/125/files#diff-4cff6710bf1e915a96f471e0a75cbbfe71dc07cf2eebabdfe07cd54878fbce14R38

**Comment by kitzberger** [comment]
@stigfaerch, i tried to debug why NewsController cannot be extended on TYPO3 10 and I seems to me that Controller-classes are not processed at all by https://github.com/georgringer/news/blob/8.6.0/Classes/Utility/ClassLoader.php#L79-L98

I'm not really familiar with the whole DI thing, but maybe it's related to this here: https://github.com/georgringer/news/pull/1512

**Comment by stigfaerch** [comment]
@kitzberger did some changes which should fix the issues mentioned. Could you have another go?

File: Classes/Utility/ClassCacheManager.php (hunk 2 above)
Comment (stigfaerch) [reply Classes/Utility/ClassCacheManager.php]: Should work now.

File: Classes/Utility/ClassCacheManager.php (hunk 1 above)
Comment (stigfaerch) [reply Classes/Utility/ClassCacheManager.php]: Now removed

**Comment by kitzberger** [comment]
@stigfaerch, with TYPO3 10 it's not working for me ;-/ The cached version of the NewsController is not being called at all ;-(

**Comment by stigfaerch** [comment]
> 



> @stigfaerch, with TYPO3 10 it's not working for me ;-/ The cached version of the NewsController is not being called at all ;-(

That's unfortunate. I have only tested this with TYPO3 11

**Comment by stigfaerch** [comment]
@georgringer Do you have time to look into this and https://github.com/georgringer/eventnews/pull/125?
I have done some of the work, hope for others to help to cross the finish line.

**Comment by georgringer** [comment]
thanks for your work!

**Comment by DavidBruchmann** [comment]
thanks a lot everybody involved!

**Pull request merged**